HT-376: Data migration to shift clips to account for new blocks introduced by HT-374

tombogle opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/hearthis/170)
<!-- Reviewable:end -->

andrew-polk reviewed (changes requested):
Would probably be good to get a second reviewer. John Thomson is out for the year.

> Reviewed 12 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170#-:-MOmhSsYABcaC01KoyjM:b5nhksw)* status: all files reviewed, 5 unresolved discussions (waiting on @tombogle)

___
*[DistFiles/FCBH Requirements to Process HearThis Recordings.pdf, line 0 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmZhZc7AKxiWJ3G6bc:-MOmZhZc7AKxiWJ3G6bd:bdqyw1) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/DistFiles/FCBH Requirements to Process HearThis Recordings.pdf#L0)):*
Are you sure you want to add this as part of this change?
If removed, don't forget the new line in .guidsForInstaller.xml.
___
*[src/HearThis/Properties/Settings.Designer.cs, line 324 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmZGnaE-5nIEbJC3eV:-MOmZGnaE-5nIEbJC3eW:b2tipx3) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Properties/Settings.Designer.cs#L324)):*
> ```
> [global::System.Configuration.DefaultSettingValueAttribute("3")]
> ```
Why didn't this get bumped?
___
*[src/HearThis/Publishing/ClipRepository.cs, line 204 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmcaj5CAUYzoxyn9HE:-MOmcaj5CAUYzoxyn9HF:b-efjeui) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L204)):*
> ```Smalltalk
> 			{
> 				if (file.Item2 > lineNumber)
> 					RobustFile.Delete(file.Item1);
> ```

You can now give your tuple fields meaningful names.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples#tuple-field-names
___
*[src/HearThis/Publishing/ClipRepository.cs, line 236 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmdNzGCfQDo4i3PXL6:-MOmdNzGCfQDo4i3PXL7:b-raa6q3) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L236)):*
> ```
> block
> ```
Should this be iLine?
You have a mixture of using "block" and line" between the comment, the method name, and this parameter name.
___
*[src/HearThis/Publishing/ClipRepository.cs, line 249 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmdsUF9NlMhqFCwDC0:-MOmdsUF9NlMhqFCwDC1:b-5yh5p4) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L249)):*
> ```Smalltalk
> 				if (firstFileShifted == -1)
> 					firstFileShifted = file.Item2;
> 				RobustFile.Move(file.Item1, Path.ChangeExtension((file.Item2 + 1).ToString(), Path.GetExtension(file.Item1)));
> ```

I guess HearThis devs would normally know such a thing, but it seems odd to me that the line number is the file extension. Or am I misinterpreting?
Might be helped by a comment / example.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[DistFiles/FCBH Requirements to Process HearThis Recordings.pdf, line  at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmZhZc7AKxiWJ3G6bc:-MOnvjTzEfspW6y7PNDL:bfgkhvz) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/DistFiles/FCBH Requirements to Process HearThis Recordings.pdf#L)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Are you sure you want to add this as part of this change?
If removed, don't forget the new line in .guidsForInstaller.xml.
</blockquote></details>

Thanks for catching that. I got trigger happy trying to push my changes right before running out to go Christmas caroling.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Properties/Settings.Designer.cs, line 324 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmZGnaE-5nIEbJC3eV:-MOnw9RL0QB6lP7wrrtU:b-csa8iu) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Properties/Settings.Designer.cs#L324)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> [global::System.Configuration.DefaultSettingValueAttribute("3")]
> ```
Why didn't this get bumped?
</blockquote></details>

I don't know. Something is broken to where Designer isn't doing this automatically, so I have to remember to do it. I'm not sure if it actually matters, but if it does, then it's really annoying to have to do it by hand.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 236 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmdNzGCfQDo4i3PXL6:-MOnyWse32x_ELk9lLvv:b-vbgap7) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L236)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> block
> ```
Should this be iLine?
You have a mixture of using "block" and line" between the comment, the method name, and this parameter name.
</blockquote></details>

I changed the method name to say Blocks. We used to use "line" in HearThis, but after deciding on the Glyssen project that "block" was less ambiguous, I started trying to use that in HearThis as well. But it's definitely not consistent throughout.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 249 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmdsUF9NlMhqFCwDC0:-MOnyzZeEIWFFWVFF9Fe:bq46h6w) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L249)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I guess HearThis devs would normally know such a thing, but it seems odd to me that the line number is the file extension. Or am I misinterpreting?
Might be helped by a comment / example.
</blockquote></details>

No, the line number is the "name." The extension is usually .wav, but in the case of a skip, it is .skip. I haven't tested any of this yet (still Draft), but I hope that Change extension will add the correct extension.


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 5 unresolved discussions (waiting on @andrew-polk)

___
*[src/HearThis/Publishing/ClipRepository.cs, line 204 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmcaj5CAUYzoxyn9HE:-MOnxfpp3s3TTd1HEiq-:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L204)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

You can now give your tuple fields meaningful names.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples#tuple-field-names
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 3 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/Publishing/ClipRepository.cs, line 236 at r1](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOmdNzGCfQDo4i3PXL6:-MOqLdaM94KQGPme64zS:b-9j2kj8) ([raw file](https://github.com/sillsdev/hearthis/blob/6d54da38fa77944f13c45274e1323f9758e64ca5/src/HearThis/Publishing/ClipRepository.cs#L236)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

I changed the method name to say Blocks. We used to use "line" in HearThis, but after deciding on the Glyssen project that "block" was less ambiguous, I started trying to use that in HearThis as well. But it's definitely not consistent throughout.
</blockquote></details>

I still think you might want to consider `iBlock`.
___
*[src/HearThis/Publishing/ClipRepository.cs, line 237 at r2](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOqLvog3GSBJX2hVZpE:-MOqLvog3GSBJX2hVZpF:bwarzde) ([raw file](https://github.com/sillsdev/hearthis/blob/222440d27f392693d33a34b93581bd1410637ed1/src/HearThis/Publishing/ClipRepository.cs#L237)):*
> ```Smalltalk
> 		// safely migrate any affected chapters.
> 		public static bool ShiftClipsAfterBlockIfAllClipsAreBeforeDate(string projectName, string bookName, int chapterNumber1Based, int block, DateTime cutoff, IScriptProvider scriptProvider)
> 		{
> ```

Using named tuple fields would greatly help this method.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 237 at r2](https://reviewable.io/reviews/sillsdev/hearthis/170#-MOqLvog3GSBJX2hVZpE:-MP4jHf28Nzl1g5z7kUq:bbhl34n) ([raw file](https://github.com/sillsdev/hearthis/blob/222440d27f392693d33a34b93581bd1410637ed1/src/HearThis/Publishing/ClipRepository.cs#L237)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Using named tuple fields would greatly help this method.
</blockquote></details>

I couldn't figure out how in this context, but I went ahead and introduced a little private class that makes it more readable and absorbs some of the boring details (and will possibly prove useful down the road). 


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 5 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/Script/ProjectSettings.cs, line 76 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4ryaN8BnniO9pcFJ1:-MP4ryaN8BnniO9pcFJ2:b-84gy58) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/Script/ProjectSettings.cs#L76)):*
> ```Smalltalk
> 		[XmlAttribute("urlForHelpWithDataMigrationProblem")]
> 		[DefaultValue("")]
> 		public string UrlForHelpWithDataMigrationProblem { get; set; }
> ```

Seems odd for this to be a project setting
___
*[src/HearThis/Script/ScriptProviderBase.cs, line 204 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4uj1o9xBENXJfSmJE:-MP4uj1o9xBENXJfSmJF:b-mjq29x) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/Script/ScriptProviderBase.cs#L204)):*
> ```
> DataMigrationReport_ht376.xml
> ```
Why this name? Could use a comment.
___
*[src/HearThis/UI/DataMigrationReportNagDlg.cs, line 20 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4s2utE6X8ojbAsTME:-MP4s2utE6X8ojbAsTMF:b71kmm7) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/UI/DataMigrationReportNagDlg.cs#L20)):*
> ```Smalltalk
> 	/// and a home for the control that allows choosing to limit the pack to the current
> 	/// actor.
> 	/// </summary>
> ```

Summary needs to be changed.
But looks like this file is a work in progress anyway
___
*[src/HearThisTests/ClipRepositoryTests.cs, line 1474 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4wbSJEW6kHE-jSOIv:-MP4wbSK7PXWjYr5ziDn:bbgi6nj) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThisTests/ClipRepositoryTests.cs#L1474)):*
> ```
> NoFilesChangedReturnsFalse
> ```
I'm not understanding something here. The SUT returns true


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 5 unresolved discussions (waiting on @andrew-polk and @tombogle)



<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Script/ProjectSettings.cs, line 76 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4ryaN8BnniO9pcFJ1:-MP5GbRKCvy_-lZdXec7:bq9b8kz) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/Script/ProjectSettings.cs#L76)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Seems odd for this to be a project setting
</blockquote></details>

Trying to plan for a possible future situation where a different problem could require a different URL. Since a project does not get migrated until it is opened, it's possible that some projects might have different problems at different points along the migration path. I'm not at all sure that my current strategy is anywhere near extensible and flexible enough to cover all the future scenarios, but I didn't really want to hardcode a single URL. Maybe a better approach is a single "token" that maps to both a filename and an entry in a table (hard-coded)  to get the URL.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Script/ScriptProviderBase.cs, line 204 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4uj1o9xBENXJfSmJE:-MP5QUKcC6SPAyfoyHU3:bae5m4a) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/Script/ScriptProviderBase.cs#L204)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> DataMigrationReport_ht376.xml
> ```
Why this name? Could use a comment.
</blockquote></details>

Decided to just use the failed version number.


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: 16 of 22 files reviewed, 5 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/UI/DataMigrationReportNagDlg.cs, line 20 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4s2utE6X8ojbAsTME:-MP5SFUNBGW17KiNBhBC:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThis/UI/DataMigrationReportNagDlg.cs#L20)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Summary needs to be changed.
But looks like this file is a work in progress anyway
</blockquote></details>

Done.
___
*[src/HearThisTests/ClipRepositoryTests.cs, line 1474 at r3](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP4wbSJEW6kHE-jSOIv:-MP5Sa_g3pGuMoClzQPu:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/5b576677c84d7b0978e7055ec391a6965507e6ee/src/HearThisTests/ClipRepositoryTests.cs#L1474)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> NoFilesChangedReturnsFalse
> ```
I'm not understanding something here. The SUT returns true
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 4 of 6 files at r4, 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: 21 of 22 files reviewed, 1 unresolved discussion (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/UI/RecordingToolControl.cs, line 1271 at r5](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP5iuvL08RJyHus8CH8:-MP5iuvL08RJyHus8CH9:b-kfikzn) ([raw file](https://github.com/sillsdev/hearthis/blob/98dbff562de5aacf2b9b6dfceca3affe8e9de00b/src/HearThis/UI/RecordingToolControl.cs#L1271)):*
> ```
> 								sourcePath = ClipPathProvider(line.Number);
> 								destPath = ClipPathProvider(destLineNumber);
> ```
I think these two lines got lost.
More precisely, these two variables now never get assigned. Should they be removed?


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 1 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 2 unresolved discussions (waiting on @tombogle)

___
*[src/HearThis/Publishing/ClipRepository.cs, line 240 at r5](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP64wqZ8ltzkCFHGnCC:-MP64wq_8-9hRtMtwg7z:b-sywqwp) ([raw file](https://github.com/sillsdev/hearthis/blob/9947d79b826f12f03e9e041874c10192153c8726/src/HearThis/Publishing/ClipRepository.cs#L240)):*
> ```
> FileName = fileName;
> ```
FilePath = filePath?


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/UI/RecordingToolControl.cs, line 1271 at r5](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP5iuvL08RJyHus8CH8:-MP6IRZM7hMF5qY-NJqC:b-5abhzk) ([raw file](https://github.com/sillsdev/hearthis/blob/98dbff562de5aacf2b9b6dfceca3affe8e9de00b/src/HearThis/UI/RecordingToolControl.cs#L1271)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> 								sourcePath = ClipPathProvider(line.Number);
> 								destPath = ClipPathProvider(destLineNumber);
> ```
I think these two lines got lost.
More precisely, these two variables now never get assigned. Should they be removed?
</blockquote></details>

Fixed. Turns out that the existing logic wasn't exactly correct either since those variables could have been null anyway. Also turns out it's really hard to test this error handling logic. 


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 240 at r5](https://reviewable.io/reviews/sillsdev/hearthis/170#-MP64wqZ8ltzkCFHGnCC:-MP6JzTS1no6YnXvOwTB:bcwau7j) ([raw file](https://github.com/sillsdev/hearthis/blob/9947d79b826f12f03e9e041874c10192153c8726/src/HearThis/Publishing/ClipRepository.cs#L240)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> FileName = fileName;
> ```
FilePath = filePath?
</blockquote></details>

I changed it, but one thing I'm sure of: throughout our code on various projects (and even in MS methods), there's abundant ambiguity between filepath and filename. There are probably even a few places where filepath is used to mean the directory containing the file, but excluding the filename itself. Unfortunate reality. 


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (commented):
> Reviewed 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
> Reviewed 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 8 unresolved discussions (waiting on @tombogle)

___
*[src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs, line 110 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAN0yn9N-cqbAkeD1b:-MPAN0yn9N-cqbAkeD1c:b-iljatk) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs#L110)):*
> ```
> Put curly braces around the text that should appear as the hyperlink.
> ```
We had some situations recently with something like this where Crowdin told the translator to be sure not to translate the text in curly brackets. I'm not sure if this is the same context or not, but it still might be worth something explicit to the translator so he knows he should translate "here", too.
Could be as simple as, `Put curly braces around the words in the translated text that should appear as a hyperlink.` Or you could explicitly say, `The word "here" should also be translated.`

Discussion: could change from curly brackets to square brackets.
___
*[src/HearThis/Utils/Utils.cs, line 65 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAMMWI84aXsndag6hq:-MPAMMWI84aXsndag6hr:b91vxni) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/Utils/Utils.cs#L65)):*
> ```Smalltalk
> 	}
> 
> 	public static class LinkLabelHelper
> ```

Probably deserves a class comment.
Also, any reason not to put it in its own file?
___
*[src/HearThis/Utils/Utils.cs, line 85 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPANmlX65CUJg3AOTyB:-MPANmlY4BRM3OUXbtlo:b-d3p9nl) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/Utils/Utils.cs#L85)):*
> ```Smalltalk
> 
> 		static List<OrigLinkLabelLinkCollection> _knownLinkLabelLinks =
> 			new List<OrigLinkLabelLinkCollection>();
> ```

Should this be private?
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 25 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAOQ9MEC0wMJiN-Wh-:-MPAOQ9MEC0wMJiN-Wh0:bzdwzll) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L25)):*
> ```
> TextWithNoCurlyBraces
> ```
TextWithInvalidCurlyBraces
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 37 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAP-lP9iWccFlx4jq5:-MPAP-lP9iWccFlx4jq6:b-tua4hv) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L37)):*
> ```Smalltalk
> 			{
> 				linkLabel.Links[0].Start = 8;
> 				linkLabel.Links[0].Length = 4;
> ```

I don't understand what this is doing. Would this be better as two tests? One for no text, one for text?

Are all these setting of start and length and linkarea before the SUT just to prove they get blown away?
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 81 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAPpi32UuC7gvofl1f:-MPAPpi32UuC7gvofl1g:bjn2f38) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L81)):*
> ```
> expectedEnd
> ```
expectedLength
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 85 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAPuPuAG5Us2UBa3ok:-MPAPuPuAG5Us2UBa3ol:b-b3za7p) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L85)):*
> ```
> thw
> ```
the
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 168 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAQweJ7-F0XSxGx-Z_:-MPAQweJ7-F0XSxGx-Za:b-w0txvx) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L168)):*
> ```
> bakc
> ```
back


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: 20 of 25 files reviewed, 8 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs, line 110 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAN0yn9N-cqbAkeD1b:-MPAcgaEDIIPp0xaSLaH:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs#L110)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> Put curly braces around the text that should appear as the hyperlink.
> ```
We had some situations recently with something like this where Crowdin told the translator to be sure not to translate the text in curly brackets. I'm not sure if this is the same context or not, but it still might be worth something explicit to the translator so he knows he should translate "here", too.
Could be as simple as, `Put curly braces around the words in the translated text that should appear as a hyperlink.` Or you could explicitly say, `The word "here" should also be translated.`

Discussion: could change from curly brackets to square brackets.
</blockquote></details>

Done.
___
*[src/HearThis/Utils/Utils.cs, line 65 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAMMWI84aXsndag6hq:-MPAqMOfDWG5ZVTrB2nC:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/Utils/Utils.cs#L65)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Probably deserves a class comment.
Also, any reason not to put it in its own file?
</blockquote></details>

Done.
___
*[src/HearThis/Utils/Utils.cs, line 85 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPANmlX65CUJg3AOTyB:-MPAqNlK9yYcI4cJR5Rj:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThis/Utils/Utils.cs#L85)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Should this be private?
</blockquote></details>

Done.
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 25 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAOQ9MEC0wMJiN-Wh-:-MPAqoJPBh15k4xsEZvu:btwsbg5) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L25)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> TextWithNoCurlyBraces
> ```
TextWithInvalidCurlyBraces
</blockquote></details>

Changed to TextWithoutPairedSquareBracketsContainingText and added a new test case.
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 37 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAP-lP9iWccFlx4jq5:-MPAsxizEpFv7YgR60Rl:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L37)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't understand what this is doing. Would this be better as two tests? One for no text, one for text?

Are all these setting of start and length and linkarea before the SUT just to prove they get blown away?
</blockquote></details>

Done.
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 81 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAPpi32UuC7gvofl1f:-MPAt3LO9HxqtCN6mwDi:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L81)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> expectedEnd
> ```
expectedLength
</blockquote></details>

Done.
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 85 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAPuPuAG5Us2UBa3ok:-MPAtMhw0gjqtSCTD0VU:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L85)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> thw
> ```
the
</blockquote></details>

Done.
___
*[src/HearThisTests/LinkLabelHelperTests.cs, line 168 at r7](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPAQweJ7-F0XSxGx-Z_:-MPAtPoQ9VjkNlfqSuN_:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/621ab65e138ac2dd51d95b37fcc546fb14214329/src/HearThisTests/LinkLabelHelperTests.cs#L168)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> bakc
> ```
back
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (changes requested):
> Reviewed 6 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 2 unresolved discussions (waiting on @tombogle)

___
*[src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs, line 111 at r8](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPB3k-S-qxAPAfSrIL7:-MPB3k-S-qxAPAfSrIL8:bkiph3c) ([raw file](https://github.com/sillsdev/hearthis/blob/cb23f3e0bb5b9c21f8d5e11941320558f8ba5702/src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs#L111)):*
> ```
> not
> ```
note
___
*[src/HearThis/Utils/Utils.cs, line 15 at r8](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPB3z4411b6v11-H9xb:-MPB3z4411b6v11-H9xc:b-qwb12l) ([raw file](https://github.com/sillsdev/hearthis/blob/cb23f3e0bb5b9c21f8d5e11941320558f8ba5702/src/HearThis/Utils/Utils.cs#L15)):*
> ```Smalltalk
> using System.Linq;
> using System.Text;
> using System.Text.RegularExpressions;
> ```

Can remove all these now


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: all files reviewed, 2 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs, line 111 at r8](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPB3k-S-qxAPAfSrIL7:-MPB6jatCllBuw-zHLL0:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/cb23f3e0bb5b9c21f8d5e11941320558f8ba5702/src/HearThis/UI/DataMigrationReportNagDlg.Designer.cs#L111)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> ```
> not
> ```
note
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andrew-polk reviewed (approved):
> Reviewed 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tombogle)



<!-- Sent from Reviewable.io -->

andrew-polk reviewed (approved):
> Reviewed 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tombogle)



<!-- Sent from Reviewable.io -->

andrew-polk reviewed (approved):
> Reviewed 4 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tombogle)



<!-- Sent from Reviewable.io -->

JohnThomson reviewed (changes requested):
Is there a problem if the user has already manually skipped a block that is now skipped by default? Will the migration make an adjustment that was already done when the block was manually skipped?

> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/170#-:-MPFTEPQ9-96do_cx-af:bkwxix3)* status: all files reviewed, 9 unresolved discussions (waiting on @tombogle)

*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFYnC0B1HBiZ9UtYUC:-MPFYnC1AOpLCno5uBV6:b-l1qsah) (no related file):*
KeepTogether seems like a new feature unrelated to the rest of the change, except that it might introduce new reasons to renumber blocks. However, renumbering blocks due to keepTogether doesn't seem to be addressed.

---
*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFZD-p8-_TInqiM8OS:-MPFZD-p8-_TInqiM8OT:b-i3h3nj) (no related file):*
Could we have tests (more integration than unit) of the process of iterating over the blocks of a chapter, considering the paragraph styles, and moving things where appropriate? As noted below, I'm wondering whether there might be interactions between the iteration code and the migration code.

---
*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFZdsuE1mTLXn9MjAv:-MPFZdsuE1mTLXn9MjAw:b-rrq1at) (no related file):*
Since I'm on vacation right now, don't wait for me to merge if you're satisfied the issues I raised are dealt with or were never real problems.

---
___
*[src/HearThis/Publishing/ClipRepository.cs, line 225 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFExPI3e8h1efGMRKQ:-MPFExPI3e8h1efGMRKR:btl8nt) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L225)):*
> ```Smalltalk
> 		}
> 
> 		private class ClipOrSkipFile : IClipFile
> ```

Might deserve class comment? I suppose "clip" is a widely used technical term in HT, but I first thought of clipping a file in the sense of removing part of it (excess silence, say). In fact, clip and skip can both be taken as verbs which can confuse. What would you think of BlockFile? The file that corresponds to a particular block, which might be its recording or an indication it was skipped.
___
*[src/HearThis/Publishing/ClipRepository.cs, line 261 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFEVVQ4FsrJmKdfqYK:-MPFEVVQ4FsrJmKdfqYL:bxlscnj) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L261)):*
> ```Smalltalk
> 			{
> 				var destPath = GetIntendedDestinationPath(positions);
> 				RobustFile.Move(FilePath, destPath);
> ```

I think this will fail if the destination already exists...probably it should have been moved first if necessary? Might be worth a comment, or even code to produce a more useful error message.
___
*[src/HearThis/Publishing/ClipRepository.cs, line 375 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFH9ng6WFSW7q6tpwJ:-MPFH9ng6WFSW7q6tpwK:b9hdyq8) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L375)):*
> ```Smalltalk
> 			// We have to move them in reverse order to avoid clobbering the next one.
> 			allFilesAfterBlock.Sort(new ClipOrSkipFileComparer(false));
> 			var lastBlockIndexToShift = allFilesAfterBlock.Last().Number;
> ```

Feels wrong. allFilesAfterBlock has been sorted so that the last (highest numbered?) block to shift is the FIRST one in it.
___
*[src/HearThis/Script/ParatextParagraph.cs, line 169 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFOu684af6-AD0h89d:-MPFOu684af6-AD0h89e:b-qrbdud) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Script/ParatextParagraph.cs#L169)):*
> ```Smalltalk
> 		///			- Handle special case of >>> (single followed by double)
> 		///		- Handle non-Roman sentence-final punctuation
> 		///		- Keep other post-end-of-sentence characters attached (closing parenthesis? >>>?
> ```

Comment should probably mention keepTogether
___
*[src/HearThis/Script/ScriptProviderBase.cs, line 199 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFRAhV7y-pqb63rDGX:-MPFRAhV7y-pqb63rDGY:bkzndge) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Script/ScriptProviderBase.cs#L199)):*
> ```Smalltalk
> 							var chaptersPotentiallyNeedingManualMigration = new Dictionary<string, List<int>>();
> 							var stopwatch = Stopwatch.StartNew();
> 							ProcessBlocksWhere(s => StylesToSkipByDefault.Contains(s.ParagraphStyle),
> ```

It's not clear to me how the implementation of ProcessBlocksWhere interacts with the migration code (here and in ChapterInfo) in cases where more than one block in the same chapter has a problem style. The chapter info code is messing with block numbers. Will that mess with the blockIndex that ProcessBlocksWhere passes to this delegate? Does migration change the number of blocks in the chapter? If so, does that affect correct termination of the loop from 0 to GetScriptBlockCount()? Is there a special case when two adjacent blocks have problem styles?
___
*[src/HearThisTests/ClipRepositoryTests.cs, line 1385 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFVHbS5fzUf3fZ05D9:-MPFVHbS5fzUf3fZ05DA:bng3yob) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThisTests/ClipRepositoryTests.cs#L1385)):*
> ```Smalltalk
> 			const int kTestChapter = 1;
> 			var clipsShiftedCalled = false;
> 			void OnClipRepositoryOnOnClipsShifted(string name, string bookName, IScriptProvider provider, int number, int clip, int @by) => clipsShiftedCalled = true;
> ```

The double "on" in this method makes me think the event should just be called 'ClipsShifted'


<!-- Sent from Reviewable.io -->

tombogle commented:
Before the change that required this migration, the \r and intro table blocks never appeared at all, so the user couldn't manually skip them.





<!-- Sent from Reviewable.io -->

tombogle commented:
*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFYnC0B1HBiZ9UtYUC:-MPHu2ja56i-efxubhCa:bsjc0ls) (no related file):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

KeepTogether seems like a new feature unrelated to the rest of the change, except that it might introduce new reasons to renumber blocks. However, renumbering blocks due to keepTogether doesn't seem to be addressed.
</blockquote></details>

It's related. I added keepTogether to allow that same method to be used to get the trimming and the creation of an IEnumerable<ScriptLine> for the special case of these new paragraph styles that should never get broken up into more than one block. This keeps things like Luk 2.3 from getting broken up into two blocks. (Probably the method should never break on a period that occurs between two numbers anyway.) Likewise, any periods that might appear in an intro table of contents row probably should not be treated as a sentence break.

---


<!-- Sent from Reviewable.io -->

tombogle commented:
*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFZdsuE1mTLXn9MjAv:-MPHuvAQ9aDeUm-_M7Kn:b8zzlkd) (no related file):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Since I'm on vacation right now, don't wait for me to merge if you're satisfied the issues I raised are dealt with or were never real problems.
</blockquote></details>

Acknowledged

---


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 225 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFExPI3e8h1efGMRKQ:-MPHv9bpEsqWKPlIVUHR:bfwu1qa) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L225)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Might deserve class comment? I suppose "clip" is a widely used technical term in HT, but I first thought of clipping a file in the sense of removing part of it (excess silence, say). In fact, clip and skip can both be taken as verbs which can confuse. What would you think of BlockFile? The file that corresponds to a particular block, which might be its recording or an indication it was skipped.
</blockquote></details>

Renamed to BlockClipOrSkipFile. BlockFile might be okay since we don't currently have any other kind of file that corresponds to a single block. But ClipOrSkip helps (for me) to keep in mind that the file that corresponds to a block could be either a clip or a "skip" file. Alternatively we could call it BlockWavOrSkipFile. But the fact that clips are recorded as WAV files is a technical detail that is not the focus of the class. 


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 261 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFEVVQ4FsrJmKdfqYK:-MPHxU6v788oMeRzgwVV:b-896fix) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L261)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think this will fail if the destination already exists...probably it should have been moved first if necessary? Might be worth a comment, or even code to produce a more useful error message.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Publishing/ClipRepository.cs, line 375 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFH9ng6WFSW7q6tpwJ:-MPHyThb0mEZq0d6y3cU:bij6by1) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Publishing/ClipRepository.cs#L375)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Feels wrong. allFilesAfterBlock has been sorted so that the last (highest numbered?) block to shift is the FIRST one in it.
</blockquote></details>

I guess it depends on whether you think of "last" as chronological or in sequential block order. I changed the name to firstBlockIndexToShift and added a comment. (This logic used to be more complex and I had a comment before, but I apparently axed it when I simplified the logic to use the comparer and sort the array in place.)


<!-- Sent from Reviewable.io -->

tombogle commented:
___
*[src/HearThis/Script/ScriptProviderBase.cs, line 199 at r11](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFRAhV7y-pqb63rDGX:-MPHzOuJ5R3LUoPztIAQ:b-xlsdkz) ([raw file](https://github.com/sillsdev/hearthis/blob/4c0f78e2a6c7505ffaeffc96f3e17ff7a4da70b8/src/HearThis/Script/ScriptProviderBase.cs#L199)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It's not clear to me how the implementation of ProcessBlocksWhere interacts with the migration code (here and in ChapterInfo) in cases where more than one block in the same chapter has a problem style. The chapter info code is messing with block numbers. Will that mess with the blockIndex that ProcessBlocksWhere passes to this delegate? Does migration change the number of blocks in the chapter? If so, does that affect correct termination of the loop from 0 to GetScriptBlockCount()? Is there a special case when two adjacent blocks have problem styles?
</blockquote></details>

Good question. As I thought about it, at first I thought there would be a special case in the 3->4 migration if there were more than one addition in succession (probably very unlikely in the case of \r markers, but quite likely in the case of the intro table styles). However, I think the logic is fine (though not as efficient as it could be) because the existing files would all get bumped up one number from the point of the first newly inserted block and then they (including the first one that had just gotten bumped) would all get bumped up again.
As for the block index in ProcessBlocksWhere and in ChapterInfo, they are different. The former is 0-based  and the latter is 1-based. They both correspond to the same blocks (represented by the sequence of blocks in the script), but the ones in ChapterInfo are informational and historical. The do not currently get updated automatically when the script changes due to text ual edits (though the idea is that we might someday find useful ways to automatically shift clips based on this information.


<!-- Sent from Reviewable.io -->

tombogle commented:
*[a discussion](https://reviewable.io/reviews/sillsdev/hearthis/170#-MPFZD-p8-_TInqiM8OS:-MPI3ofX8n1Cs4_FwE4C:bou45yz) (no related file):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Could we have tests (more integration than unit) of the process of iterating over the blocks of a chapter, considering the paragraph styles, and moving things where appropriate? As noted below, I'm wondering whether there might be interactions between the iteration code and the migration code.
</blockquote></details>

Probably not a bad idea, but I don't know if it's worth postponing the fix to do this. After addressing all your concerns, I didn't find any actual logic errors in the code. I should probably think about this in the morning when I'm more alert.  

---


<!-- Sent from Reviewable.io -->
